Bolt Sampler - Neo4j 4.x and cluster features

pecollet opened the pull request:
## Description

Bolt Sampler : 
- added "database" option, required for Neo4j 4.x multi-database support
- added "access mode" option, that allows running against a Neo4j Enterprise causal cluster, with proper query routing using the "neo4://" or "bolt+routing://" URL schemes.
- added "transaction timeout" option, which is a useful parameter when load testing
- fixed error displaying results when "Record Query Results" is enabled, with 4.x Driver (retrieving resultSummary caused early consumption of results).

## Motivation and Context
Enables running JMeter against Neo4j 4.0, 4.1 & 4.2.
Enables running JMeter against Neo4j Enterprise Causal clusters
Enables setting a limit on request execution time.

## How Has This Been Tested?
Tested against : 3.5 standalone Neo4j instance, 3.5 Causal Cluster, 4.2 standalone Neo4j instance, 4.2 Causal Cluster.
Test config : 1 "Bolt Connection Configuration" + 1 "Bolt Request" in a Thread group

## Screenshots (if appropriate):
<img width="1427" alt="Jmeter-bolt-sampler" src="https://user-images.githubusercontent.com/45878183/102241396-d154c980-3ef0-11eb-8043-0c444ebd1217.png">

## Types of changes
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the [code style][style-guide] of this project.
- [x] I have updated the documentation accordingly.

pecollet committed:
added database and access mode parameters to BoltSampler

database : required to address a on-default dtabase in Neo4j 4.x
accessMode : allow to test against a cluster w/o having all queries sent to the leader (previous was defaulting to WRITE)

pecollet committed:
Added transaction timeout option

- new tx timeout option
- fixed results consumption issue in 4.x
- separated options from query in UI + better descriptions

pecollet committed:
Delete Bolt Request.jmx

pecollet committed:
styling fixes & edited changes file

pecollet committed:
Unit tests

FSchumacher commented on src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/AbstractBoltTestElement.java:
```diff
@@ -48,4 +84,27 @@ public boolean isRecordQueryResults() {
+        SessionConfig.Builder sessionConfigBuilder = SessionConfig.builder()
+                .withDefaultAccessMode(Enum.valueOf(AccessMode.class, getAccessMode()));
+
+        if (database != null && !"".equals(database)) {
```
You could probably use `StringUtils#isNotBlank(String)` to make intention more clear.

FSchumacher commented on src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/BoltSampler.java:
```diff
@@ -135,12 +146,25 @@ private String request() {
+            //get records already as consume() will exhaust the stream
+            records = result.list();
+        } else {
+            records = null;
```
Would an empty List work here, too? (`Collections#emptyList()`) That way, we would not have to deal with a null pointer later.

FSchumacher commented on src/protocol/bolt/src/main/resources/org/apache/jmeter/protocol/bolt/sampler/BoltSamplerResources.properties:
```diff
@@ -17,9 +17,16 @@
+accessMode.displayName=Access Mode
+accessMode.shortDescription=Whether it's a READ or WRITE query (affects query routing in clusters)
+database.displayName=Database
+database.shortDescription=Neo4j 4.x : database to query (leave empty for 3.5)
```
There should not be a space in front of the colon in the English translations. You can certainly add it for the French translation :)

FSchumacher commented on src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/BoltTestElementBeanInfoSupport.java:
```diff
@@ -45,5 +47,23 @@ protected BoltTestElementBeanInfoSupport(Class<? extends TestBean> beanClass) {
+    }
+
+    private String[] getListAccessModes() {
+        String[] list = {AccessMode.READ.toString(), AccessMode.WRITE.toString()};
```
Could be returned directly.

FSchumacher commented on src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/AbstractBoltTestElement.java:
```diff
@@ -17,13 +17,49 @@
+    }
+
+    public void setAccessMode(String accessMode) {
+        this.accessMode = accessMode;
```
Should we check this against the available access modes from the driver?

FSchumacher reviewed (commented):
Thanks for the PR. I think it will have to wait until the release 5.4.1, which should happen soon.

pecollet committed:
changes based on feedback

also removed time assertions that made some unit tests flaky

pecollet committed:
Merge pull request #1 from apache/master

update to latest JMeter

pecollet committed:
Connection Pool Max Size parameter added to Bolt Connection Config

Allows to specify a pool size larger than the default 100, which can become a bottleneck with large number of threads using a single Neo4j driver.
Also improve Bolt URI description to describe neo4j:// format.

pecollet committed:
Merge branch 'master' of https://github.com/pecollet/jmeter

ham1 commented on src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/AbstractBoltTestElement.java:
```diff
@@ -17,13 +17,53 @@
+    }
+
+    public String getAccessMode() {
+        if (accessMode != null) {
```
Consider writing this using a guard clause? i.e.
```
if (accessMode == null) {
    return AccessMode.WRITE.toString();
}
return accessMode;
```

pecollet committed:
getAccessMode with guard clause

also updated the docs of the Bolt Sampler and Bolt config to reflect the new parameters

pecollet committed:
Merge branch 'master' into master

pecollet committed:
formatting fixes in changes.xml

pecollet committed:
Merge branch 'master' of https://github.com/pecollet/jmeter

pecollet committed:
more xml formatting fixes in changes.xml

pecollet committed:
Merge branch 'master' into master

FSchumacher commented:
Thanks for the PR. Can you test the next nightly or build from trunk?
